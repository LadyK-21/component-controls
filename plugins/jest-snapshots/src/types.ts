export const renderers = {
  rtl: 'react-testing-library',
  rtr: 'react-test-renderer',
  enzyme: 'enzyme-react-17',
};

export type Renderers = keyof typeof renderers;

export type TeplateFormats = 'cjs' | 'esm' | 'ts';
