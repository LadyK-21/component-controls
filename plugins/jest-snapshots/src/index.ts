import { loadStore } from '@component-controls/store';
import { RendererFn } from '@component-controls/test-renderers';
import { getBundleName } from '@component-controls/webpack-compile';

export type RendererFnResult = Promise<any> | any;

export const runJestSnapshots = (
  renderer: RendererFn,
  bundleName?: string,
): void => {
  const bundle = bundleName || getBundleName();
  const store = loadStore(require(bundle));
  const config = store.config;
  Object.keys(store.docs).forEach(docId => {
    const doc = store.docs[docId];

    if (doc.stories && doc.stories.length) {
      const stories = doc.stories;
      describe(doc.title, () => {
        stories.forEach(storyId => {
          const story = store.stories[storyId];
          it(story.name, async () => {
            const { toJson } = (await renderer({ story, doc, config })) || {};
            expect(toJson ? toJson() : undefined).toMatchSnapshot();
          });
        });
      });
    }
  });
};
