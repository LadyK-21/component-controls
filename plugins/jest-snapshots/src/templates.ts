import fs from 'fs';
import path from 'path';
import dot from 'dot';
import { defaultConfigFolder } from '@component-controls/core/node-utils';
import { Renderers, TeplateFormats, renderers } from './types';

dot.templateSettings.strip = false;
(dot as any).log = false;

export type TemplateOptions = {
  renderer?: Renderers;
  format?: TeplateFormats;
  configPath?: string;
  bundle?: string;
  out?: string;
  name?: string;
};
export const createTemplate = (options?: TemplateOptions): string => {
  const {
    renderer = 'rtl',
    format = 'cjs',
    configPath: config = path.resolve(process.cwd(), defaultConfigFolder),
    name = 'component-controls generated',
    out,
    bundle,
  } = options || {};
  const filePath = path.resolve(
    __dirname,
    `../templates/all-store.${format}.js`,
  );

  const importsPath = path.resolve(
    __dirname,
    `../templates/imports/${renderers[renderer]}.${format}.js`,
  );
  const store = bundle ? 'bundle' : 'imports';
  const storeImportPath = path.resolve(
    __dirname,
    `../templates/store-import/${store}.${format}.js`,
  );
  const storeLoopPath = path.resolve(
    __dirname,
    `../templates/store-loop/${store}.${format}.js`,
  );
  const storeRenderPath = path.resolve(
    __dirname,
    `../templates/store-render/${store}.${format}.js`,
  );
  const storeRender = fs.readFileSync(storeRenderPath, 'utf8');
  const renderPath = path.resolve(
    __dirname,
    `../templates/render/${renderers[renderer]}.${format}.js`,
  );
  const render = dot.template(fs.readFileSync(renderPath, 'utf8'))({
    storeRender,
  });
  const bundlePath =
    out && bundle ? `.${path.sep}${path.relative(out, bundle)}` : bundle;
  const configPath = out ? path.relative(out, config) : config;
  const vars = {
    imports: fs.readFileSync(importsPath, 'utf8'),
    render,
    configPath,
    storeRender,
    storeImports: fs.readFileSync(storeImportPath, 'utf8'),
    bundlePath,
    storeLoop: dot.template(fs.readFileSync(storeLoopPath, 'utf8'))({
      render,
      bundlePath,
      configPath,
    }),
    name,
  };
  const content = fs.readFileSync(filePath, 'utf8');
  const template = dot.template(content)(vars);
  return template;
};
