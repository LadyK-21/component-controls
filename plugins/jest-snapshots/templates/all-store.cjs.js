const path = require('path');
const MatchMediaMock = require('jest-matchmedia-mock').default;
{{=it.storeImports}}

{{=it.imports}}

let matchMedia;

describe('{{=it.name}}', () => {
  beforeAll(() => {
    jest.mock('rc-util/lib/Portal');
    matchMedia = new MatchMediaMock();
  });
  afterEach(() => {
    matchMedia.clear();
  });
{{=it.storeLoop}}
  
});
