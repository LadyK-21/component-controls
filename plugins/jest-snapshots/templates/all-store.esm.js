import * as path from 'path';
import MatchMediaMock from 'jest-matchmedia-mock';
{{=it.storeImports}}

{{=it.imports}}

let matchMedia;

describe('{{=it.name}}', () => {
  beforeAll(() => {
    jest.mock('rc-util/lib/Portal');
    matchMedia = new MatchMediaMock();
  });
  afterEach(() => {
    matchMedia.clear();
  });
{{=it.storeLoop}}  
});
