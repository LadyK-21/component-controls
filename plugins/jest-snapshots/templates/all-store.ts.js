import path from 'path';
import MatchMediaMock from 'jest-matchmedia-mock';
{{=it.storeImports}}

{{=it.imports}}

let matchMedia: MatchMediaMock;

describe('{{=it.name}}', () => {
  beforeAll(() => {
    jest.mock('rc-util/lib/Portal');
    matchMedia = new MatchMediaMock();
  });
  afterEach(() => {
    matchMedia.clear();
  });
{{=it.storeLoop}} 
});
