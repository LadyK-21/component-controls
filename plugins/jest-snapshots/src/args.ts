import { ArgOptions } from '@component-controls/webpack-compile/cli';
import { renderers } from '@component-controls/test-renderers';

export const jestCliArgs: ArgOptions = [
  {
    name: 'renderer',
    options: {
      alias: 'r',
      description: 'jest framework renderer',
      choices: Object.keys(renderers),
      default: 'react',
    },
  },
  {
    name: 'test',
    options: {
      alias: 't',
      description: 'tests output folder',
      type: 'string',
      default: 'tests',
    },
  },
];
