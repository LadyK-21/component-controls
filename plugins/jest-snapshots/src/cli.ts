import path from 'path';
import fs from 'fs';
import { cliArgs } from '@component-controls/webpack-compile/cli';
import { jestCliArgs } from './args';
import { createTemplate } from './templates';

export const run = async (): Promise<void> => {
  const args = cliArgs(jestCliArgs);
  const parsedArgs = args.parse();
  let testFolder =
    (parsedArgs.output as string) || path.resolve(process.cwd(), 'tests');

  if (!path.isAbsolute(testFolder)) {
    testFolder = path.resolve(process.cwd(), testFolder);
  }

  if (!fs.existsSync(testFolder)) {
    fs.mkdirSync(testFolder);
  }
  const {
    renderer,
    format,
    overwrite,
    config,
    test = 'stories.test.js',
    bundle,
    name,
  } = parsedArgs;
  const testFilePath = path.resolve(testFolder, test);

  if (!overwrite && fs.existsSync(testFilePath)) {
    console.warn(
      `${testFilePath} already exists and you have chosen not to overwrite it.`,
    );
    return;
  }
  const content = createTemplate({
    renderer,
    format,
    configPath: config,
    name,
    bundle,
  });
  fs.writeFileSync(testFilePath, content, 'utf8');
};
