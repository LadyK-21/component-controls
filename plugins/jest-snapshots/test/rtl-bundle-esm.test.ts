import path from 'path';
import { createTemplate } from '../src/templates';
import { runTests } from './run-tests';

jest.setTimeout(100000);
test('rtl bundle esm', async () => {
  const renderedFile = createTemplate({
    format: 'esm',
    renderer: 'rtl',
    bundle: path.resolve(__dirname, 'bundle/component-controls.js'),
  });
  expect(renderedFile).toMatchSnapshot();
  await runTests('esm', renderedFile);
});
