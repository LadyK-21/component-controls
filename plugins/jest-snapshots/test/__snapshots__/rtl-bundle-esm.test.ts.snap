// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rtl bundle esm 1`] = `
"import * as path from 'path';
import MatchMediaMock from 'jest-matchmedia-mock';
import { loadStore } from '@component-controls/store';
const { render: reactRender } = require('@component-controls/render/react');

import { render, act } from '@testing-library/react';


let matchMedia;

describe('component-controls generated', () => {
  beforeAll(() => {
    jest.mock('rc-util/lib/Portal');
    matchMedia = new MatchMediaMock();
  });
  afterEach(() => {
    matchMedia.clear();
  });
const store = loadStore(require(path.resolve(__dirname, './bundle/component-controls.js')));
const renderFn = store.config.renderFn || reactRender;
Object.keys(store.docs).forEach(docId => {
  const doc = store.docs[docId];
  if (doc.stories && doc.stories.length) {
    const stories = doc.stories;
    describe(doc.title, () => {
      stories.forEach(storyId => {
        const story = store.stories[storyId];
        it(story.name, async () => {
              let rendered;
              act(() => {
              rendered = renderFn(story, doc);
              });
              let serialize;
              if (rendered) {
                ({ asFragment: serialize } = render(rendered));
              } else {
                serialize = () => 'error';
              }

          expect(serialize()).toMatchSnapshot();
        });
      });
    });
  }
});  
});
"
`;
